cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -56,6 +45,21 @@ const (
 
 	certManagerImageComponent = "cert-manager"
 	timeoutConfigKey          = "cert-manager-timeout"
+
+	certmanagerVersionAnnotation = "certmanager.clusterctl.cluster.x-k8s.io/version"
+	certmanagerHashAnnotation    = "certmanager.clusterctl.cluster.x-k8s.io/hash"
+
+	// NOTE: // If the cert-manager.yaml asset is modified, this line **MUST** be updated
+	// accordingly else future upgrades of the cert-manager component will not
+	// be possible, as there'll be no record of the version installed.
+	embeddedCertManagerManifestVersion = "v0.16.0"
+
+	// NOTE: The hash is used to ensure that when the cert-manager.yaml file is updated,
+	// the version number marker here is _also_ updated.
+	// You can either generate the SHA256 hash of the file, or alternatively
+	// run `go test` against this package. THe Test_VersionMarkerUpToDate will output
+	// the expected hash if it does not match the hash here.
+	embeddedCertManagerManifestHash = "5770f5f01c10a902355b3522b8ce44508ebb6ec88955efde9a443afe5b3969d7"
```

Comment:
Should we be hardcoding versions and hashes within the codebase? Can we open an issue to get rid of this or find an alternative that doesn't require code changes?

Reply:
This was inherited from https://github.com/kubernetes-sigs/cluster-api/pull/3313 and slightly re-worked here.
So far we have a unit test that ensures this value is in sync with the embedded manifest.

But happy to open an issue for another round of improvement

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/3505